src/lib/Libutil/range.c

```diff
@@ -475,72 +500,131 @@ range_remove_value(range **r, int val)
  *
  */
 int
-range_add_value(range *r, int val, enum range_step_type type)
+range_add_value(range **r, int val, int range_step)
 {
 	range *cur;			/* current range structure in list */
-	range *prev = NULL;		/* previous range structure in list */
-	range *new;			/* if we need to add to the list */
-	int done = 0;			/* have we added val to our range, leave loop */
+	range *next_range;
 
 	if (r == NULL)
 		return 0;
 
-	if (range_contains(r, val))
+	if (*r == NULL) {
+		/* If there are no range structs in the list; create the new range for first value */
+		range * first_range = NULL;
+		if ((first_range = new_range()) == NULL) {
+			return 0;
+		}
+		first_range->count = 1;
+		first_range->step = range_step;
```

Comment:
Since most of the range.c utility functions depend on step which resides in each element we need to maintain it. 

It is achievable in server-side to call these range functions with "step factor" by referring job array obj, but in the scheduler case it is not feasible since there is no commonplace to refer the "step factor", Thus looking into each range struct for any computation like "delete", "find next", "parse_range".

In Scheduler, one way to make all range structs independent of step is by fetching "array_indices_submitted"(it has step input) along with current attributes (https://github.com/openpbs/openpbs/blob/master/src/scheduler/job_info.c#L942) from the server.

To make this struct independent of step element, 
1) "Step" should be stored at job object in Scheduler, 
2) Need to refactor all places in Server, Scheduler, Simulator to pass this "step" in all range functions.

Reply:
ok i think we can live with it. I do not see the range DS used anywhere else other than array jobs, in which case the parent job has a step that covers for all the ranges. We could save a bit of memory by not duplicating the same value again and again. I see the argument that this might be used in the future, but we can always change things when we actually need in the future. 

For now, assuming the number of ranges are really small, and that the scheduler does not have a parent job object (really?) I think we are okay